From 9923cf65deaf3421ebeece3556c4e7b69459496f Mon Sep 17 00:00:00 2001
From: Christopher Albert <albert@tugraz.at>
Date: Mon, 22 Dec 2025 13:42:13 +0100
Subject: [PATCH] gimplify: map Fortran DT scalars for OpenACC enter data

When OpenACC enter data maps only an allocatable component, gimplify creates a
struct mapping for the containing derived type.  Ensure scalar non-pointer
fields of the derived type are also copied to the device so later kernels can
read them reliably.

For Fortran descriptors, keep the struct member list in ascending field order
and do not split descriptor-related nodes that follow the member list.

Also initialize a few locals to avoid -Wmaybe-uninitialized warnings observed
during local builds.

	PR fortran/123252
	PR fortran/123255

gcc/ChangeLog:

	* gimplify.cc (omp_accumulate_sibling_list): For OpenACC enter data in
	Fortran, add map(to:) clauses for scalar fields and insert them into the
	GOMP_MAP_STRUCT member list in field order.
	* omp-low.cc (lower_omp_target): Initialize labels used for optional
	Fortran descriptor handling.

gcc/testsuite/ChangeLog:

	* gfortran.dg/goacc/pr123252-scalar-field-mapping.f90: New test.
	* gfortran.dg/goacc/pr123255-allocatable-component-map-order.f90: New test.

Signed-off-by: Christopher Albert <albert@tugraz.at>
---
 gcc/gimplify.cc                               | 160 ++++++++++++++++--
 gcc/omp-low.cc                                |   9 +-
 .../goacc/pr123252-scalar-field-mapping.f90   |  37 ++++
 ...123255-allocatable-component-map-order.f90 |  25 +++
 4 files changed, 213 insertions(+), 18 deletions(-)
 create mode 100644 gcc/testsuite/gfortran.dg/goacc/pr123252-scalar-field-mapping.f90
 create mode 100644 gcc/testsuite/gfortran.dg/goacc/pr123255-allocatable-component-map-order.f90

diff --git a/gcc/gimplify.cc b/gcc/gimplify.cc
index 95506f1682f..cf63225c04a 100644
--- a/gcc/gimplify.cc
+++ b/gcc/gimplify.cc
@@ -12693,6 +12693,17 @@ omp_accumulate_sibling_list (enum omp_region_type region_type,
 
       struct_map_to_clause->put (base, l);
 
+      tree scalar_first = NULL_TREE;
+      tree *scalar_tail = &scalar_first;
+      int scalar_count = 0;
+      bool scalar_copy = false;
+      if (!openmp && code == OACC_ENTER_DATA && lang_GNU_Fortran ())
+	{
+	  enum gomp_map_kind kind = OMP_CLAUSE_MAP_KIND (*grp_start_p);
+	  if (kind == GOMP_MAP_TO_PSET || (kind & GOMP_MAP_FLAG_TO) != 0)
+	    scalar_copy = true;
+	}
+
       /* On first iterating through the clause list, we insert the struct node
 	 just before the component access node that triggers the initial
 	 omp_accumulate_sibling_list call for a particular sibling list (and
@@ -12710,11 +12721,11 @@ omp_accumulate_sibling_list (enum omp_region_type region_type,
 	  tree sc = *insert_node_pos;
 	  OMP_CLAUSE_CHAIN (l) = desc;
 	  OMP_CLAUSE_CHAIN (*grp_start_p) = OMP_CLAUSE_CHAIN (desc);
-	  OMP_CLAUSE_CHAIN (desc) = sc;
-	  *insert_node_pos = l;
-	}
-      else if (attach_detach)
-	{
+		  OMP_CLAUSE_CHAIN (desc) = sc;
+		  *insert_node_pos = l;
+		}
+	      else if (attach_detach)
+		{
 	  tree extra_node;
 	  tree alloc_node
 	    = build_omp_struct_comp_nodes (code, *grp_start_p, grp_end,
@@ -12786,10 +12797,10 @@ omp_accumulate_sibling_list (enum omp_region_type region_type,
 			(TREE_OPERAND (adecl, 0)))) == POINTER_TYPE))
 		OMP_CLAUSE_DECL (alloc_node) = TREE_OPERAND (adecl, 0);
 
-	      *added_tail = tail;
-	    }
-	}
-      else
+		      *added_tail = tail;
+		    }
+		}
+	      else
 	{
 	  gcc_assert (*grp_start_p == grp_end);
 	  if (reprocessing_struct)
@@ -12814,14 +12825,133 @@ omp_accumulate_sibling_list (enum omp_region_type region_type,
 	      *insert_node_pos = l;
 	      *added_tail = &OMP_CLAUSE_CHAIN (alloc_node);
 	    }
-	  else
-	    grp_start_p = omp_siblist_insert_node_after (l, insert_node_pos);
+		  else
+		    grp_start_p = omp_siblist_insert_node_after (l, insert_node_pos);
+		}
+
+      if (scalar_copy)
+	{
+	  tree type = TREE_TYPE (base);
+	  if (TREE_CODE (type) == REFERENCE_TYPE)
+	    type = TREE_TYPE (type);
+
+	  tree *member_start_p = &OMP_CLAUSE_CHAIN (l);
+	  while (*member_start_p
+		 && (OMP_CLAUSE_MAP_KIND (*member_start_p)
+		     == GOMP_MAP_FIRSTPRIVATE_POINTER
+		     || OMP_CLAUSE_MAP_KIND (*member_start_p)
+			== GOMP_MAP_FIRSTPRIVATE_REFERENCE))
+	    member_start_p = &OMP_CLAUSE_CHAIN (*member_start_p);
+
+	  unsigned HOST_WIDE_INT n_members = tree_to_uhwi (OMP_CLAUSE_SIZE (l));
+
+	  auto field_bytepos = [] (tree clause) -> HOST_WIDE_INT
+	  {
+	    tree d = OMP_CLAUSE_DECL (clause);
+	    if (TREE_CODE (d) == COMPONENT_REF)
+	      return int_byte_position (TREE_OPERAND (d, 1));
+	    return HOST_WIDE_INT_MAX;
+	  };
+
+	  auto member_has_field = [&] (tree field) -> bool
+	  {
+	    tree member = *member_start_p;
+	    for (unsigned HOST_WIDE_INT i = 0; i < n_members && member;
+		 i++, member = OMP_CLAUSE_CHAIN (member))
+	      {
+		tree mdecl = OMP_CLAUSE_DECL (member);
+		if (TREE_CODE (mdecl) == COMPONENT_REF
+		    && TREE_OPERAND (mdecl, 1) == field)
+		  return true;
+	      }
+	    return false;
+	  };
+
+	  if (RECORD_OR_UNION_TYPE_P (type))
+	    for (tree field = TYPE_FIELDS (type); field;
+		 field = DECL_CHAIN (field))
+	      {
+		if (TREE_CODE (field) != FIELD_DECL
+		    || DECL_ARTIFICIAL (field))
+		  continue;
+
+		tree ftype = TREE_TYPE (field);
+		if (POINTER_TYPE_P (ftype)
+		    || TREE_CODE (ftype) == ARRAY_TYPE
+		    || RECORD_OR_UNION_TYPE_P (ftype))
+		  continue;
+
+		if (!INTEGRAL_TYPE_P (ftype)
+		    && !SCALAR_FLOAT_TYPE_P (ftype))
+		  continue;
+
+		if (!TYPE_SIZE_UNIT (ftype)
+		    || TREE_CODE (TYPE_SIZE_UNIT (ftype)) != INTEGER_CST)
+		  continue;
+
+		if (member_has_field (field))
+		  continue;
+
+		tree cref = build3_loc (OMP_CLAUSE_LOCATION (grp_end),
+					COMPONENT_REF, ftype,
+					unshare_expr (base), field, NULL_TREE);
+		tree c2 = build_omp_clause (OMP_CLAUSE_LOCATION (grp_end),
+					    OMP_CLAUSE_MAP);
+		OMP_CLAUSE_SET_MAP_KIND (c2, GOMP_MAP_TO);
+		OMP_CLAUSE_DECL (c2) = cref;
+		OMP_CLAUSE_SIZE (c2) = TYPE_SIZE_UNIT (ftype);
+		OMP_CLAUSE_CHAIN (c2) = NULL_TREE;
+		*scalar_tail = c2;
+		scalar_tail = &OMP_CLAUSE_CHAIN (c2);
+		scalar_count++;
+	      }
+
+	  if (scalar_first)
+	    {
+	      tree *extra_start_p = member_start_p;
+	      for (unsigned HOST_WIDE_INT i = 0;
+		   i < n_members && *extra_start_p; i++)
+		extra_start_p = &OMP_CLAUSE_CHAIN (*extra_start_p);
+
+	      tree extra = *extra_start_p;
+	      *extra_start_p = NULL_TREE;
+
+	      for (tree scalar = scalar_first; scalar; )
+		{
+		  tree next = OMP_CLAUSE_CHAIN (scalar);
+		  HOST_WIDE_INT pos = field_bytepos (scalar);
+
+		  tree *ins = member_start_p;
+		  for (unsigned HOST_WIDE_INT i = 0; i < n_members && *ins; i++)
+		    {
+		      if (field_bytepos (*ins) > pos)
+			break;
+		      ins = &OMP_CLAUSE_CHAIN (*ins);
+		    }
+
+		  OMP_CLAUSE_CHAIN (scalar) = *ins;
+		  *ins = scalar;
+		  n_members++;
+
+		  scalar = next;
+		}
+
+	      extra_start_p = member_start_p;
+	      for (unsigned HOST_WIDE_INT i = 0;
+		   i < n_members && *extra_start_p; i++)
+		extra_start_p = &OMP_CLAUSE_CHAIN (*extra_start_p);
+	      *extra_start_p = extra;
+
+	      OMP_CLAUSE_SIZE (l)
+		= size_binop (PLUS_EXPR, OMP_CLAUSE_SIZE (l),
+			      size_int (scalar_count));
+	    }
 	}
 
-      unsigned last_access = base_token + 1;
+	      unsigned last_access = base_token + 1;
 
-      while (last_access + 1 < addr_tokens.length ()
-	     && addr_tokens[last_access + 1]->type == ACCESS_METHOD)
+	      while (last_access + 1 < addr_tokens.length ()
+		     && addr_tokens[last_access + 1]->type == ACCESS_METHOD)
 	last_access++;
 
       if ((region_type & ORT_TARGET)
@@ -18408,7 +18538,7 @@ gimplify_omp_loop (tree *expr_p, gimple_seq *pre_p)
 
   TREE_SET_CODE (for_stmt, OMP_SIMD);
 
-  int last;
+  int last = 0;
   switch (kind)
     {
     case OMP_CLAUSE_BIND_THREAD: last = 0; break;
diff --git a/gcc/omp-low.cc b/gcc/omp-low.cc
index 4540a25d1ad..0689ef94440 100644
--- a/gcc/omp-low.cc
+++ b/gcc/omp-low.cc
@@ -13661,10 +13661,13 @@ lower_omp_target (gimple_stmt_iterator *gsi_p, omp_context *ctx)
 	    if (OMP_CLAUSE_CODE (c) != OMP_CLAUSE_HAS_DEVICE_ADDR
 		&& lang_hooks.decls.omp_array_data (ovar, true))
 	      {
-		tree not_null_lb, null_lb, after_lb;
+		tree not_null_lb = NULL_TREE;
+		tree null_lb = NULL_TREE;
+		tree after_lb = NULL_TREE;
 		tree var1, var2, size1, size2;
 		tree present = omp_check_optional_argument (ovar, true);
-		if (present)
+		bool has_present = present != NULL_TREE;
+		if (has_present)
 		  {
 		    location_t clause_loc = OMP_CLAUSE_LOCATION (c);
 		    not_null_lb = create_artificial_label (clause_loc);
@@ -13688,7 +13691,7 @@ lower_omp_target (gimple_stmt_iterator *gsi_p, omp_context *ctx)
 		size2 = fold_build2 (POINTER_DIFF_EXPR, ssizetype,
 				   build_fold_addr_expr (var1), var);
 		size2 = fold_convert (sizetype, size2);
-		if (present)
+		if (has_present)
 		  {
 		    tree tmp = create_tmp_var (TREE_TYPE (var1));
 		    gimplify_assign (tmp, var1, &ilist);
diff --git a/gcc/testsuite/gfortran.dg/goacc/pr123252-scalar-field-mapping.f90 b/gcc/testsuite/gfortran.dg/goacc/pr123252-scalar-field-mapping.f90
new file mode 100644
index 00000000000..01424529241
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/goacc/pr123252-scalar-field-mapping.f90
@@ -0,0 +1,37 @@
+! { dg-do compile }
+! { dg-options "-O2 -fopenacc -fdump-tree-omplower-details" }
+
+program pr123252
+  use, intrinsic :: iso_fortran_env, only: dp => real64
+  implicit none
+
+  type :: container_t
+    real(dp), allocatable :: arr(:)
+    logical :: flag
+    integer :: n
+  end type container_t
+
+  type(container_t) :: c
+  integer :: i
+
+  c%n = 8
+  c%flag = .true.
+  allocate(c%arr(c%n))
+  do i = 1, c%n
+    c%arr(i) = real(i, dp)
+  end do
+
+  !$acc enter data copyin(c%arr(1:c%n))
+
+  !$acc parallel loop present(c%arr)
+  do i = 1, c%n
+    if (c%flag) then
+      c%arr(i) = c%arr(i) + 1.0_dp
+    end if
+  end do
+  !$acc end parallel loop
+end program pr123252
+
+! { dg-final { scan-tree-dump "oacc_enter_data" "omplower" } }
+! { dg-final { scan-tree-dump "map\\(to:c\\.flag" "omplower" } }
+! { dg-final { scan-tree-dump "map\\(to:c\\.n" "omplower" } }
diff --git a/gcc/testsuite/gfortran.dg/goacc/pr123255-allocatable-component-map-order.f90 b/gcc/testsuite/gfortran.dg/goacc/pr123255-allocatable-component-map-order.f90
new file mode 100644
index 00000000000..c1d38199dd7
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/goacc/pr123255-allocatable-component-map-order.f90
@@ -0,0 +1,25 @@
+! { dg-do compile }
+! { dg-options "-O2 -fopenacc -fdump-tree-omplower" }
+!
+! PR fortran/123255
+! Verify that mapping a later allocatable component also maps earlier scalar
+! components in field order without splitting descriptor-related nodes.
+
+program pr123255
+  implicit none
+
+  type :: container_t
+    integer :: n
+    real(8), allocatable :: data(:)
+  end type container_t
+
+  type(container_t) :: c
+
+  c%n = 2
+  allocate (c%data(c%n))
+
+  !$acc enter data copyin(c%data)
+  !$acc exit data delete(c%data)
+end program pr123255
+
+! { dg-final { scan-tree-dump-times "oacc_enter_data.*map\\\\(to:c\\\\.n \\\\[len: \\\\[0-9\\\\]+\\\\]\\\\).*map\\\\(to:c\\\\.data \\\\[pointer set, len: \\\\[0-9\\\\]+\\\\]\\\\)" 1 "omplower" } }
-- 
2.52.0

